server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -575,8 +592,18 @@ static int getTopDocsSize(SearchRequest request) {
          * Creates a new search response from the given merged hits.
          * @see #merge(boolean, ReducedQueryPhase, Collection, IntFunction)
          */
-        public InternalSearchResponse buildResponse(SearchHits hits) {
-            return new InternalSearchResponse(hits, aggregations, suggest, shardResults, timedOut, terminatedEarly, numReducePhases);
+        public InternalSearchResponse buildResponse(SearchHits hits, Collection<? extends SearchPhaseResult> fetchResults) {
+            SearchProfileResults profileResults = mergeProfile(fetchResults);
```

Comment:
I'm sort of worried that adding too much merging logic to this class'd make it harder to think about because it's super big and has different threads coming at it. I really just pulled this out to make the null guard more obvious. I'd be ok doing whatever you think is right here.